pstoreds/addr_book.go

```diff
@@ -2,398 +2,608 @@ package pstoreds
-	ttl  time.Duration
-	addr ma.Multiaddr
+// addrsRecord decorates the AddrBookRecord with locks and metadata.
+type addrsRecord struct {
+	sync.RWMutex
+	*pb.AddrBookRecord
+	dirty bool
 }
 
-func (ar *addrRecord) MarshalBinary() ([]byte, error) {
-	ttlB := make([]byte, 8)
-	binary.LittleEndian.PutUint64(ttlB, uint64(ar.ttl))
-	return append(ttlB, ar.addr.Bytes()...), nil
+func (r *addrsRecord) Reset() {
+	r.AddrBookRecord.Reset()
+	r.dirty = false
 }
 
-func (ar *addrRecord) UnmarshalBinary(b []byte) error {
-	ar.ttl = time.Duration(binary.LittleEndian.Uint64(b))
-	// this had been serialized by us, no need to check for errors
-	ar.addr, _ = ma.NewMultiaddrBytes(b[8:])
+// FlushInTxn writes the record to the datastore by calling ds.Put, unless the record is
+// marked for deletion, in which case the deletion is executed via ds.Delete.
+func (r *addrsRecord) FlushInTxn(txn ds.Txn) (err error) {
+	key := addrBookBase.ChildString(b32.RawStdEncoding.EncodeToString([]byte(r.Id.ID)))
+	if len(r.Addrs) == 0 {
+		return txn.Delete(key)
+	}
+	// cannot use a buffer pool because data is retained in the txn until it's committed or discarded.
+	data, err := r.Marshal()
+	if err != nil {
+		return err
+	}
+	if err = txn.Put(key, data); err != nil {
+		return err
+	}
+	// write succeeded; record is no longer dirty.
+	r.dirty = false
 	return nil
 }
 
-// NewAddrBook initializes a new address book given a
-// Datastore instance, a context for managing the TTL manager,
-// and the interval at which the TTL manager should sweep the Datastore.
-func NewAddrBook(ctx context.Context, store ds.TxnDatastore, opts Options) (*dsAddrBook, error) {
-	if _, ok := store.(ds.TTLDatastore); !ok {
-		return nil, ErrTTLDatastore
+// Flush creates a ds.Txn, and calls FlushInTxn with it.
+func (r *addrsRecord) Flush(ds ds.TxnDatastore) (err error) {
+	txn, err := ds.NewTransaction(false)
+	if err != nil {
+		return err
 	}
+	defer txn.Discard()
 
-	var (
-		cache cache = &noopCache{}
-		err   error
-	)
+	if err = r.FlushInTxn(txn); err != nil {
+		return err
+	}
+	return txn.Commit()
+}
 
-	if opts.CacheSize > 0 {
-		if cache, err = lru.NewARC(int(opts.CacheSize)); err != nil {
-			return nil, err
+// Refresh is called on records to perform housekeeping. The return value signals if the record was changed
+// as a result of the refresh.
+//
+// Refresh does the following:
+// * sorts the addresses by expiration (soonest expiring first).
+// * removes the addresses that have expired.
+//
+// It short-circuits optimistically when we know there's nothing to do.
+//
+// Refresh is called from several points:
+// * when accessing and loading an entry.
+// * when performing periodic GC.
+// * after an entry has been modified (e.g. addresses have been added or removed,
+//   TTLs updated, etc.)
+//
+// If the return value is true, the caller can perform a flush immediately, or can schedule an async
+// flush, depending on the context.
+func (r *addrsRecord) Refresh() (chgd bool) {
```

Comment:
Nit: I'd call this something like `Clean`. To me, `Refresh` means "reload from disk".

Reply:
Good point.